### PR TITLE
Alekseev football tournament

### PR DESCRIPTION
**Pull request by YarComicSans**
Fully reworked how the programm works. Before I used Lists to remember which players play and what matches are played at the current round.
Now I use trees for this purpose
I use a list of trees to know what matches and which players play and each grid. For example - in Double elumination I have a tree for players with no loses and a tree for players with no loses. This way it would be easier to operate with matches and players, and it wouldn't be a problem to add new grid trees (Triple elumination, etc...)
I have two classes for drawing grids - one for Vertical drawing and one for Horizontal drawing. They need to be configured a little still since I have a problem with end-game (when we have more than two grids how will the end-game will work)
For saving/loading progress I have two XML files - each for their own type of tournament (single/double elumination). The progress is saved once a new grid is created, new round is played or on exit if the users desides so. Loading requires user to create at least one grid (that will create XML files). After that the user can load the progress whenever feels like it.

### CODE REVIEW

File: 2018/Alekseev/FootballLeague/Football_League/Program.cs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            _currentMatchesGrid2 = ChooseWinnerOfMatchesInGrid(_currentMatchesGrid2);
+
+            List<Contestant> winners = new List<Contestant>();
+            List<Contestant> winnersInLosers = new List<Contestant>();
+
+            foreach (Match match in _currentMatchesGrid2)
+                winnersInLosers.Add(match.Winner);
+            if (_currentPlayersGrid2.Count % 2 != 0)
+                winnersInLosers.Add(_currentPlayersGrid2[_currentPlayersGrid2.Count - 1]);
+
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/Program.cs]: Use empty lines to separate cycles and conditions.

**Comment by YarComicSans** [comment]
Fully reworked how the programm works. Before I used Lists to remember which players play and what matches are played at the current round.
Now I use trees for this purpose
I use a list of trees to know what matches and which players play and each grid. For example - in Double elumination I have a tree for players with no loses and a tree for players with no loses. This way it would be easier to operate with matches and players, and it wouldn't be a problem to add new grid trees (Triple elumination, etc...)
I have two classes for drawing grids - one for Vertical drawing and one for Horizontal drawing. They need to be configured a little still since I have a problem with end-game (when we have more than two grids how will the end-game will work)
For saving/loading progress I have two XML files - each for their own type of tournament (single/double elumination). The progress is saved once a new grid is created, new round is played or on exit if the users desides so. Loading requires user to create at least one grid (that will create XML files). After that the user can load the progress whenever feels like it.

File: 2018/Alekseev/FootballLeague/Football_League/Match.cs [hunk 2]
```diff
@@ -1,46 +1,47 @@
...
+            else
             {
-                Loser = null;
-                Winner = _playerTwo;
-                return true;
+                Winner = PlayerTwo;
+                Loser = PlayerOne;
             }
-            return false;
+            return Winner;
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/Match.cs]: Don't think we need a return of Winner there. This value is stored in a public field anyway. Maybe it's not convinient to use down there, IDK.

File: 2018/Alekseev/FootballLeague/Football_League/Match.cs [hunk 3]
```diff
@@ -1,46 +1,47 @@
...
+
+        public Contestant GetLoser()
         {
-            if (AutoWin())
-                return;
-            Winner = (number == 1) ? _playerOne : _playerTwo;
-            Loser = (number == 1) ? _playerTwo : _playerOne;
-            if(_playerTwo != null)
-                Winner.Position = (_playerOne.Position + _playerTwo.Position) / 2;
+            return Loser;
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/Match.cs]: Unnecessary method

File: 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+            var currentMatch = StartMatch;
+            StartMatch = currentMatch;
+            CurrentRoundFirstMatch = currentMatch;
+            while (currentMatch.NextRoundMatch != null)
+                currentMatch = currentMatch.NextRoundMatch;
+            CurrentRoundFirstMatch = currentMatch;
+
+            var losers = new List<Contestant>();
+            //var currentMatch = CurrentRoundFirstMatch;
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs]: Forgot a comment there

File: 2018/Alekseev/FootballLeague/Football_League/Program.cs [hunk 5]
```diff
@@ -69,30 +99,20 @@ public static int RunChoice()
...
-            {
-                var newPlayer = new Contestant(contestant.Name, _scoreboard.CurrentPlayerPosition);
-                _scoreboard.CurrentPlayerPosition += newPlayer.Name.Length + 1;
-                newPlayersRandomized.Add(newPlayer);
-            }
+        public static List<Contestant> AddPlayersToCompetition(int num)
+        {
+            var players = new List<Contestant>();
+            for (int i = 0; i < num; i++)
+                players.Add(new Contestant(ConsoleWorker.GetPlayerName()));
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/Program.cs]: Please, use empty lines and braces

File: 2018/Alekseev/FootballLeague/Football_League/HorizontalDrawer.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                    _horizontalGrid.Add("");
+                foreach (var str in result)
+                {
+                    _horizontalGrid.Add(str);
+                }
+            }
+        }
+
+        public List<string> MakeGridTree(MatchTreeGrid grid)
+        {
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/HorizontalDrawer.cs]: Probably it's a too large method. Need to split

File: 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            StartMatch = CurrentRoundFirstMatch;
+        }
+        public List<Contestant> PlayRound()
+        { // check if startMatchNextRoundMatch exists ...
+            if(CurrentRoundFirstMatch == null)
+                return new List<Contestant>();
+
+            var currentMatch = StartMatch;
+            StartMatch = currentMatch;
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs]: 2 equal lines

File: 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            CurrentRoundFirstMatch = currentMatch;
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/MatchTreeGrid.cs]: This value is changed again 2 lines down. This line is unnecesssary.

File: 2018/Alekseev/FootballLeague/Football_League/VerticalDrawer.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                {
+                    if (_verticalGrid.Count <= j + i * 3)
+                        _verticalGrid.Add(result[j]);
+                    else
+                        _verticalGrid[j + i * 3] += result[j];
+                }
+            }
+        }
+
+        public List<string> MakeGridTree(MatchTreeGrid gridTree)
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/VerticalDrawer.cs]: Again, a big method. Need to split

File: 2018/Alekseev/FootballLeague/Football_League/ConsoleWorker.cs [hunk 10]
```diff
@@ -73,177 +73,231 @@ public static int ChooseDrawingType()
         public static void PrintHorizontalGrid(List<string> horizontalGrid)
```
Comment (taler0n) [inline 2018/Alekseev/FootballLeague/Football_League/ConsoleWorker.cs]: Probably, it's better to move drawers to distinct classes. They have quite many methods processing them.

**Pull request merged**